File: packages/sdkwork-utils-php/src/ObjectUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class ObjectUtils
{
    public static function pick(array $source, array $keys): array
    {
        $result = [];
        foreach ($keys as $key) {
            if (array_key_exists($key, $source)) {
                $result[$key] = $source[$key];
            }
        }
        return $result;
    }

    public static function omit(array $source, array $keys): array
    {
        $blocked = array_flip($keys);
        $result = [];
        foreach ($source as $key => $value) {
            if (!isset($blocked[$key])) {
                $result[$key] = $value;
            }
        }
        return $result;
    }

    public static function getPath(array $source, string $path): mixed
    {
        $current = $source;
        foreach (self::splitPath($path) as $segment) {
            if (!is_array($current) || !array_key_exists($segment, $current)) {
                return null;
            }
            $current = $current[$segment];
        }
        return $current;
    }

    public static function hasPath(array $source, string $path): bool
    {
        $value = self::getPath($source, $path);
        return $value !== null;
    }

    public static function shallowMerge(array $base, array $overlay): array
    {
        return array_merge($base, $overlay);
    }

    public static function setPath(array $source, string $path, mixed $value): array
    {
        $segments = self::splitPath($path);
        if ($segments === []) {
            return $source;
        }
        $current = &$source;
        foreach (array_slice($segments, 0, -1) as $segment) {
            if (!isset($current[$segment]) || !is_array($current[$segment])) {
                $current[$segment] = [];
            }
            $current = &$current[$segment];
        }
        $current[$segments[array_key_last($segments)]] = $value;
        return $source;
    }

    public static function deepMerge(array $base, array $overlay): array
    {
        $result = $base;
        foreach ($overlay as $key => $value) {
            if (isset($result[$key]) && is_array($result[$key]) && is_array($value)) {
                $result[$key] = self::deepMerge($result[$key], $value);
            } else {
                $result[$key] = $value;
            }
        }
        return $result;
    }

    public static function compact(array $source): array
    {
        return array_filter($source, static fn ($value) => $value !== null);
    }

    public static function keys(array $source): array
    {
        return array_keys($source);
    }

    public static function values(array $source): array
    {
        return array_values($source);
    }

    private static function splitPath(string $path): array
    {
        return array_values(array_filter(explode('.', $path), static fn ($part) => $part !== ''));
    }
}

File: packages/sdkwork-utils-csharp/ObjectUtils.cs
namespace Sdkwork.Utils;

public static class ObjectUtils
{
    public static Dictionary<string, object?> Pick(Dictionary<string, object?> source, IEnumerable<string> keys)
    {
        var result = new Dictionary<string, object?>();
        foreach (var key in keys)
        {
            if (source.TryGetValue(key, out var value)) result[key] = value;
        }
        return result;
    }

    public static Dictionary<string, object?> Omit(Dictionary<string, object?> source, IEnumerable<string> keys)
    {
        var blocked = keys.ToHashSet();
        return source.Where(entry => !blocked.Contains(entry.Key)).ToDictionary(entry => entry.Key, entry => entry.Value);
    }

    public static object? GetPath(object? source, string path)
    {
        object? current = source;
        foreach (var segment in SplitPath(path))
        {
            if (current is not Dictionary<string, object?> map || !map.TryGetValue(segment, out current)) return null;
        }
        return current;
    }

    public static bool HasPath(object? source, string path)
    {
        var value = GetPath(source, path);
        return value is not null;
    }

    public static Dictionary<string, object?> ShallowMerge(
        Dictionary<string, object?> baseValue,
        Dictionary<string, object?> overlay)
    {
        var result = new Dictionary<string, object?>(baseValue);
        foreach (var (key, value) in overlay)
        {
            result[key] = value;
        }
        return result;
    }

    public static Dictionary<string, object?> SetPath(Dictionary<string, object?> source, string path, object? value)
    {
        var segments = SplitPath(path);
        if (segments.Count == 0) return source;

        var root = new Dictionary<string, object?>(source);
        var current = root;
        for (var index = 0; index < segments.Count - 1; index++)
        {
            var segment = segments[index];
            if (current.TryGetValue(segment, out var next) && next is Dictionary<string, object?> nextMap)
            {
                nextMap = new Dictionary<string, object?>(nextMap);
                current[segment] = nextMap;
                current = nextMap;
            }
            else
            {
                var created = new Dictionary<string, object?>();
                current[segment] = created;
                current = created;
            }
        }
        current[segments[^1]] = value;
        return root;
    }

    public static object? DeepMerge(object? baseValue, object? overlay)
    {
        if (baseValue is Dictionary<string, object?> baseMap && overlay is Dictionary<string, object?> overlayMap)
        {
            var result = new Dictionary<string, object?>(baseMap);
            foreach (var (key, overlayValue) in overlayMap)
            {
                result[key] = result.TryGetValue(key, out var existing) ? DeepMerge(existing, overlayValue) : overlayValue;
            }
            return result;
        }
        return overlay;
    }

    public static Dictionary<string, object?> Compact(Dictionary<string, object?> source) =>
        source.Where(entry => entry.Value is not null).ToDictionary(entry => entry.Key, entry => entry.Value);

    public static List<string> Keys(Dictionary<string, object?> source) => source.Keys.ToList();

    public static List<object?> Values(Dictionary<string, object?> source) => source.Values.ToList();

    private static List<string> SplitPath(string path) =>
        path.Split('.', StringSplitOptions.RemoveEmptyEntries).ToList();
}
